smithy-aws-cloudformation-traits/src/main/java/software/amazon/smithy/aws/cloudformation/traits/ResourceIndex.java

```diff
@@ -0,0 +1,479 @@
+import software.amazon.smithy.utils.MapUtils;
+import software.amazon.smithy.utils.SetUtils;
+
+/**
+ * Index of resources to their CloudFormation identifiers
+ * and properties.
+ *
+ * <p>This index performs no validation that the identifiers
+ * and reference valid shapes.
+ */
+public final class ResourceIndex implements KnowledgeIndex {
+
+    static final Set<Mutability> FULLY_MUTABLE = SetUtils.of(
+            Mutability.CREATE, Mutability.READ, Mutability.WRITE);
+
+    private final Model model;
+    private final Map<ShapeId, Map<String, ResourcePropertyDefinition>> resourcePropertyMutabilities = new HashMap<>();
+    private final Map<ShapeId, Set<ShapeId>> resourceExcludedProperties = new HashMap<>();
+    private final Map<ShapeId, Set<String>> resourcePrimaryIdentifiers = new HashMap<>();
+    private final Map<ShapeId, List<Set<String>>> resourceAdditionalIdentifiers = new HashMap<>();
+
+    /**
+     * CloudFormation-specific property mutability options.
+     */
+    public enum Mutability {
+        CREATE,
+        READ,
+        WRITE
+    }
+
+    public ResourceIndex(Model model) {
+        this.model = model;
+
+        OperationIndex operationIndex = OperationIndex.of(model);
+        model.shapes(ResourceShape.class)
+                .flatMap(shape -> Trait.flatMapStream(shape, ResourceTrait.class))
+                .forEach(pair -> {
+                    ResourceShape resource = pair.getLeft();
+                    ShapeId resourceId = resource.getId();
+
+                    // Start with the explicit resource identifiers.
+                    resourcePrimaryIdentifiers.put(resourceId, SetUtils.copyOf(resource.getIdentifiers().keySet()));
+                    setIdentifierMutabilities(resource);
+
+                    // Use the read lifecycle's input to collect the additional identifiers
+                    // and its output to collect readable properties.
+                    resource.getRead().ifPresent(operationId -> {
+                        operationIndex.getInput(operationId).ifPresent(input -> {
+                            addAdditionalIdentifiers(resource, computeResourceAdditionalIdentifiers(input));
+                        });
+                        operationIndex.getOutput(operationId).ifPresent(output -> {
+                            updatePropertyMutabilities(resourceId, operationId, output,
+                                    SetUtils.of(Mutability.READ), this::addReadMutability);
+                        });
+                    });
+
+                    // Use the put lifecycle's input to collect put-able properties.
+                    resource.getPut().ifPresent(operationId -> {
+                        operationIndex.getInput(operationId).ifPresent(input -> {
+                            updatePropertyMutabilities(resourceId, operationId, input,
+                                    SetUtils.of(Mutability.CREATE, Mutability.WRITE), this::addPutMutability);
```

Comment:
Ah, so we do automatically infer the mutability characteristics of properties based on which lifecycle the property is associated with.